File: src/main/java/com/softwareplace/encrypt/Encrypt.java
package com.softwareplace.encrypt;

import java.util.Random;

import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;

public class Encrypt {
	public static final int ENCODER_STRENGTH = 5;
	public static final BCryptPasswordEncoder PASSWORD_ENCODER = new BCryptPasswordEncoder(ENCODER_STRENGTH);

	private final String password;
	private final String encodedPassword;
	private final String token;
	private final String authToken;
	private final String salt;

	public Encrypt(String password) {
		this.password = password;
		this.encodedPassword = PASSWORD_ENCODER.encode(password);
		this.token = PASSWORD_ENCODER.encode(encodedPassword + mixedString());
		this.authToken = PASSWORD_ENCODER.encode(token);
		this.salt = PASSWORD_ENCODER.encode(this.authToken + this.token);
	}

	private String mixedString() {
		int randomResult = new Random().nextInt(100_000);
		return String.valueOf(System.currentTimeMillis() + randomResult);
	}

	public boolean isValidPassword(String encodedPassword) {
		return PASSWORD_ENCODER.matches(this.password, encodedPassword);
	}

	public String getEncodedPassword() {
		return encodedPassword;
	}

	public String getToken() {
		return token;
	}

	public String getAuthToken() {
		return authToken;
	}

	public String getSalt() {
		return salt;
	}
}

File: src/main/kotlin/com/softwareplace/springsecurity/encrypt/Encrypt.kt
package com.softwareplace.springsecurity.encrypt

import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import java.util.*

@Suppress("JoinDeclarationAndAssignment", "MemberVisibilityCanBePrivate")
class Encrypt(private val password: String) {

    val encodedPassword: String
    val token: String
    val authToken: String
    val salt: String

    init {
        encodedPassword = PASSWORD_ENCODER.encode(password)
        token = PASSWORD_ENCODER.encode(encodedPassword + mixedString())
        authToken = PASSWORD_ENCODER.encode(token)
        salt = PASSWORD_ENCODER.encode(authToken + token)
    }

    private fun mixedString(): String {
        val randomResult = Random().nextInt(100000)
        return (System.currentTimeMillis() + randomResult).toString()
    }

    fun isValidPassword(encodedPassword: String?): Boolean {
        return PASSWORD_ENCODER.matches(password, encodedPassword)
    }

    companion object {
        const val ENCODER_STRENGTH = 5
        val PASSWORD_ENCODER = BCryptPasswordEncoder(ENCODER_STRENGTH)
    }
}
